x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportGetUserPrivilegesAction.java

```diff
@@ -42,12 +45,25 @@ public TransportGetUserPrivilegesAction(
     @Override
     protected void doExecute(Task task, GetUserPrivilegesRequest request, ActionListener<GetUserPrivilegesResponse> listener) {
         securityContext.requireUser();
-        final Subject subject = securityContext.getAuthentication().getEffectiveSubject();
+        final Authentication authentication = securityContext.getAuthentication();
+        final Subject subject = authentication.getEffectiveSubject();
         if (subject.getUser().principal().equals(request.username()) == false) {
             listener.onFailure(new IllegalArgumentException("users may only list the privileges of their own account"));
             return;
         }
 
-        authorizationService.retrieveUserPrivileges(subject, securityContext.getAuthorizationInfoFromContext(), listener);
+        try {
+            authorizationService.retrieveUserPrivileges(subject, securityContext.getAuthorizationInfoFromContext(), listener);
+        } catch (UnsupportedOperationException e) {
```

Comment:
I decided to use `try-catch` instead of pro-actively checking the Role class because:
1. Whether LimitedRole can throw exception on building privileges response is an implementation detail. The LimitedRole class is in a separate module (`core`) and in theory does not even need to be exposed to the security module. In fact, security module currently has no direct reference to it.
2. Checking the Role class does not really work with Custom Authorization engine by checking because it is theoretically possible for a custom engine to implement the `retrieveUserPrivileges` call in a different way.

Reply:
My only reservation here is that we are leaking this implementation detail all the way up to the transport layer in an unstructured way: we are assuming that any `UnsupportedOperationException` thrown by `retrieveUserPrivileges` for an API key is because our implementation of `retrieveUserPrivileges` doesn't support this. `UnsupportedOperationException` is generic though so this could lead to unexpected behavior/incorrect error messages for custom implementations (or if we throw `UnsupportedOperationException` somewhere else throughout `retrieveUserPrivileges`).

I'm wondering if we should specialize the exception thrown in `LimitedRole` to a sublass of `UnsupportedOperationException`. This way we still leak the implementation but in a more robust, structured way. 

Alternatively, we could throw the 400 API-key specific exception inside `RBACEngine` which would keep it closer to the root cause and not force our interpretation on exceptions thrown by custom implementations.

WDYT?

Reply:
I agree it's awkward. The problem is not every API key should fail. Only those with assigned role descriptors will fail. Underlying those API keys have their roles represented as `LimitedRole`. But I don't want to expose `LimiteRole` to the security module because it should just rely on the `Role` interface. I pondered on this and didn't find a clear better solution:

1. Throwing specific subclass of `UnsupportedOperationException` (or just a specific exception) is slightly better. But it still leaks implementation. I also didn't want to add another class just for this error message. My thought was that the exception + checking authentication being an API key (`authentication.isApiKey()` inside the catch block) is sufficient to rule out accidental unexpected behaviour in future. 
2. By "throwing 400 API key specific exception inside RBACEngine", I assume you mean actively checking whether the `Role` is a `LimitedRole` and throw error accordingly? This would require `RBACEngine` be aware of `LimitedRole` which I rather dislike as stated above.
3. If you mean use try-catch for `UnsupportedOperationException` in RBACEngine, this means putting the error message inside RBACEngine. I initially dislike this because I felt this error message, especially with the HTTP url should be a handle level thing. But I now noticed that similar messages are prepared at other service classes (mostly implementations of the new `HealthIndicatorService`). So maybe this is not too bad.

So overall, I think option 3 is my current preference. More concretely, catching `UnsupportedOperationException` in `RBACEngine#getUserPrivileges` and retrow it as a `IllegalArgumentException` with proper error message. Using `IllegalArgumentException` avoids having to set an explicit 400 status code which is something I also didn't like at service layer. Please let me know if this works for you.

Reply:
Exactly, I meant option 3 👍